File: python/sdk/calls.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Dict, Optional
import os
import requests


@dataclass
class ApiResponse:
    status_code: int
    headers: Dict[str, str]
    data: Any


@dataclass
class CallCreateResponse:
    status: str
    call_id: str
    message: str


class CallsClient:
    BASE_URL = "http://localhost:3005"

    def __init__(self, api_key: str, base_url: Optional[str] = None, timeout: Optional[float] = 30.0,
                 default_organization_id: Optional[str] = None):
        if not api_key:
            raise ValueError("API key is required")
        self.api_key = api_key
        self.base_url = (base_url or self.BASE_URL).rstrip("/")
        self.timeout = timeout
        # Prefer explicit default_organization_id; otherwise fall back to env
        self.default_org_id = default_organization_id or os.getenv("NEXA_ORGANIZATION_ID")
        self._base_headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }

    def _headers(self, organization_id: Optional[str] = None) -> Dict[str, str]:
        headers = dict(self._base_headers)
        if organization_id:
            headers["Organization-Id"] = organization_id
        return headers

    def list_calls(
        self,
        organization_id: Optional[str] = None,
        *,
        agent_id: Optional[str] = None,
        call_id: Optional[str] = None,
        batch_call_id: Optional[str] = None,
        from_: Optional[str] = None,
    ) -> Any:
        # Resolve org: explicit > default set on client > env
        org = organization_id or self.default_org_id or os.getenv("NEXA_ORGANIZATION_ID")
        if not org:
            raise ValueError("organization_id is required")
        params = {
            "agent_id": agent_id,
            "call_id": call_id,
            "batch_call_id": batch_call_id,
            "from": from_,
        }
        # remove Nones
        params = {k: v for k, v in params.items() if v is not None and str(v) != ""}
        headers = self._headers(org)
        # Send both header variants for compatibility
        headers.setdefault("X-Organization-Id", headers.get("Organization-Id"))
        resp = requests.get(f"{self.base_url}/calls", headers=headers, params=params or None, timeout=self.timeout)
        data: Any
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Calls list failed ({resp.status_code}): {data}")
        return data

    def create_call(
        self,
        *,
        organization_id: Optional[str] = None,
        phone_number: str,
        agent_id: str,
        metadata: Dict[str, Any],
        agent_version_number: Optional[int] = None,
    ) -> Any:
        # Resolve org: explicit > default set on client > env
        organization_id = organization_id or self.default_org_id or os.getenv("NEXA_ORGANIZATION_ID")
        if not organization_id:
            raise ValueError("organization_id is required")
        if not phone_number:
            raise ValueError("phone_number is required")
        if not agent_id:
            raise ValueError("agent_id is required")
        if metadata is None:
            raise ValueError("metadata is required")

        body: Dict[str, Any] = {
            "phone_number": phone_number,
            "agent_id": agent_id,
            "metadata": metadata,
        }
        if agent_version_number is not None:
            body["agent_version_number"] = agent_version_number

        headers = {**self._headers(organization_id), "Content-Type": "application/json"}
        headers.setdefault("X-Organization-Id", headers.get("Organization-Id"))
        resp = requests.post(
            f"{self.base_url}/calls",
            headers=headers,
            json=body,
            timeout=self.timeout,
        )
        parsed: Any
        try:
            parsed = resp.json()
        except Exception:
            parsed = resp.text
        if not resp.ok:
            raise RuntimeError(f"Create call failed ({resp.status_code}): {parsed}")
        if isinstance(parsed, dict) and all(k in parsed for k in ("status", "call_id", "message")):
            return CallCreateResponse(status=str(parsed["status"]), call_id=str(parsed["call_id"]), message=str(parsed["message"]))
        return parsed

    def get_call_by_id(self, id: str) -> Any:
        if not id:
            raise ValueError("id is required")
        resp = requests.get(f"{self.base_url}/calls/{id}", headers=self._headers(), timeout=self.timeout)
        data: Any
        try:
            data = resp.json()
        except Exception:
            data = resp.text
        if not resp.ok:
            raise RuntimeError(f"Get call failed ({resp.status_code}): {data}")
        return data

File: js/src/calls.ts
import fetch from "cross-fetch";

const DEFAULT_BASE_URL = "http://localhost:3005";

// Methods return only the parsed response body (no statusCode/headers)

export type ClientConfig = {
    apiKey: string;
    baseUrl?: string;
    timeoutMs?: number;
};

export type ListCallsParams = {
    organizationId?: string;
    agent_id?: string;
    call_id?: string;
    batch_call_id?: string;
    from?: string;
};

export type CreateCallBody = {
    phone_number: string;
    agent_id: string;
    agent_version_number?: number;
    metadata: Record<string, any>;
};

export type CallCreateResponse = {
    status: string;
    call_id: string;
    message: string;
};

function buildUrl(baseUrl: string, path: string, params?: Record<string, string | undefined>): string {
    const url = new URL(path.startsWith("/") ? path : `/${path}`, baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`);
    if (params) {
        for (const [k, v] of Object.entries(params)) {
            if (typeof v !== "undefined" && v !== null && String(v).length > 0) {
                url.searchParams.append(k, String(v));
            }
        }
    }
    return url.toString();
}

function headersToObject(headers: any): Record<string, string> {
    const out: Record<string, string> = {};
    headers.forEach((v: string, k: string) => {
        out[k] = v;
    });
    return out;
}

export async function listCalls(
    {
        apiKey,
        baseUrl = DEFAULT_BASE_URL,
        timeoutMs,
    }: ClientConfig,
    params: ListCallsParams
): Promise<any> {
    // Directly read from env if not provided
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const orgId = process.env.NEXA_ORGANIZATION_ID;
    if (!orgId) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(
            buildUrl(baseUrl, "/calls", {
                agent_id: params.agent_id,
                call_id: params.call_id,
                batch_call_id: params.batch_call_id,
                from: params.from,
            }),
            {
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    Accept: "application/json",
                    "Organization-Id": orgId,
                    "X-Organization-Id": orgId,
                },
                signal: controller?.signal,
            }
        );

        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Calls list failed (${response.status}): ${text}`);
        }

        return data;
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function createCall(
    {
        apiKey,
        baseUrl = DEFAULT_BASE_URL,
        timeoutMs,
    }: ClientConfig,
    organizationId: string,
    body: CreateCallBody
): Promise<CallCreateResponse> {
    // Directly read from env if not provided
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const orgId = organizationId || (typeof process !== "undefined" ? (process as any).env?.NEXA_ORGANIZATION_ID : undefined);
    if (!orgId) throw new Error("organizationId is required (set NEXA_ORGANIZATION_ID or pass organizationId)");
    if (!body?.phone_number) throw new Error("phone_number is required");
    if (!body?.agent_id) throw new Error("agent_id is required");
    if (!body?.metadata) throw new Error("metadata is required");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, "/calls"), {
            method: "POST",
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
                "Content-Type": "application/json",
                "Organization-Id": orgId,
                "X-Organization-Id": orgId,
            },
            body: JSON.stringify(body),
            signal: controller?.signal,
        });

        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Create call failed (${response.status}): ${text}`);
        }
        return data as CallCreateResponse;
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function getCallById(
    {
        apiKey,
        baseUrl = DEFAULT_BASE_URL,
        timeoutMs,
    }: ClientConfig,
    id: string
): Promise<any> {
    if (!id) throw new Error("id is required");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, `/calls/${encodeURIComponent(id)}`), {
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
            },
            signal: controller?.signal,
        });

        const text = await response.text();
        let data: any;
        try { data = text ? JSON.parse(text) : null; } catch { data = text; }
        if (!response.ok) {
            throw new Error(`Get call failed (${response.status}): ${text}`);
        }
        return data;
    } finally {
        if (timer) clearTimeout(timer);
    }
}
